### PR TITLE
Add an option to upload existing image in Client Profile

### LINKED ISSUE #1182: No options to upload existing photo in Edit client page.
1) Go to left navigation page and tap on Clients.
 2) In clients list page, click on '+' icon.
 3) In next create client page, tap on image and then select take photo.
 4) Submit client with other mandatory fields.
5) Go to the client page, and tap on the image.
Expected: It should show both 'Take photo' and Upload Photo options.
Actual: Its only showing take photo option. Uploading existing photo is not  available.

### PR DESCRIPTION
**Pull request by laxyapahuja**
Fixes #1182 

<a href="https://codein.withgoogle.com/tasks/4734270094966784/?sp-organization=5721189306073088&sp-search=client">Task Link</a>

ezgif-4-25bd4b5d4d62

- [x] Apply the `MifosStyle.xml` style template to your code in Android Studio.

- [x] Run the unit tests with `./gradlew check` to make sure you didn't break anything

- [x] If you have multiple commits please combine them into one commit by squashing them.

### CODE REVIEW

**Comment by laxyapahuja** [comment]
Will fix style violations tomorrow.

**Comment by moksh-mahajan** [comment]
Hi @laxyapahuja, did you claim the task on the GCI portal before sending the PR?

**Comment by laxyapahuja** [comment]
> Hi @laxyapahuja, did you claim the task on the GCI portal before sending the PR?

Hey @moksh-mahajan , I didn't know earlier that you weren't allowed to make a PR without claiming a task. Will claim this task right after my current one gets approved.

**Comment by moksh-mahajan** [comment]
@laxyapahuja Another student has already claimed the task. So, you can claim it if he/she abandons the task or in case fails to do it in the given time frame.

**Comment by laxyapahuja** [comment]
@moksh-mahajan i've claimed this task on the dashboard now. can you please review it? :D

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java [hunk 1]
```diff
@@ -299,6 +338,27 @@ public void captureClientImage() {
...
 
+    private void checkPermissions() {
+        if (ContextCompat.checkSelfPermission(getActivity(),
+                Manifest.permission.WRITE_EXTERNAL_STORAGE)
+                != PackageManager.PERMISSION_GRANTED) {
+            ActivityCompat.requestPermissions(getActivity(),
+                    new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE},
+                    1001);
+        }   else {
+            int i = 1;
```
Comment (luckyman20) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java]: Do we need this?

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java [hunk 2]
```diff
@@ -226,8 +238,59 @@ public void onAttach(Activity activity) {
...
+                        selectedImage,
+                        filePathColumn, null, null, null);
+                // Move to first row
+                cursor.moveToFirst();
+
+                int columnIndex = cursor.getColumnIndex(filePathColumn[0]);
+                imgDecodableString = cursor.getString(columnIndex);
+                cursor.close();
+                Bitmap pickedImage = BitmapFactory.decodeFile(imgDecodableString);
+                saveBitmap("client_image", pickedImage);
```
Comment (jsahil730) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java]:
```suggestion
                saveBitmap(pickedImage);
```

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java [hunk 3]
```diff
@@ -226,8 +238,59 @@ public void onAttach(Activity activity) {
...
+                Toast.makeText(getActivity(), R.string.havent_picked_image,
+                        Toast.LENGTH_LONG).show();
+            }
+        } catch (Exception e) {
+            Toast.makeText(getActivity(), e.toString(), Toast.LENGTH_LONG)
+                    .show();
+        }
+    }
+
+    public static void saveBitmap(String bitName,
```
Comment (jsahil730) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java]:
```suggestion
    public void saveBitmap(
```

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java [hunk 4]
```diff
@@ -226,8 +238,59 @@ public void onAttach(Activity activity) {
...
+                                  Bitmap mBitmap) {
+
+        File f = new File(Environment.getExternalStorageDirectory()
```
Comment (jsahil730) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java]:
```suggestion
```

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java [hunk 5]
```diff
@@ -226,8 +238,59 @@ public void onAttach(Activity activity) {
...
+                .toString() + "/" + bitName + ".png");
```
Comment (jsahil730) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java]:
```suggestion
```

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java [hunk 6]
```diff
@@ -226,8 +238,59 @@ public void onAttach(Activity activity) {
...
+        try {
+            f.createNewFile();
```
Comment (jsahil730) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java]:
```suggestion
            ClientImageFile.createNewFile();
```

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java [hunk 7]
```diff
@@ -226,8 +238,59 @@ public void onAttach(Activity activity) {
...
+            FileOutputStream fOut = null;
+            fOut = new FileOutputStream(f);
```
Comment (jsahil730) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java]:
```suggestion
            fOut = new FileOutputStream(ClientImageFile);
```

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java (hunk 3 above)
Comment (jsahil730) [reply mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java]: Also, please remove the newline here, after making the changes @laxyapahuja

**Comment by jsahil730** [comment]
@luckyman20 I think it probably has issues when uploaded image sizes are very high, like 3-4 MB, could you take a look into this?

**Comment by laxyapahuja** [comment]
> @luckyman20 I think it probably has issues when uploaded image sizes are very high, like 3-4 MB, could you take a look into this?

It's most probably a server side problem as it showed me the error as 403 HTTP FORBIDDEN.

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java [hunk 8]
```diff
@@ -226,8 +237,55 @@ public void onAttach(Activity activity) {
...
+                Toast.makeText(getActivity(), R.string.havent_picked_image,
+                        Toast.LENGTH_LONG).show();
+            }
+        } catch (Exception e) {
+            Toast.makeText(getActivity(), e.toString(), Toast.LENGTH_LONG)
+                    .show();
+        }
+    }
+
+    public void saveBitmap(File f, Bitmap mBitmap) {
```
Comment (luckyman20) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java]:
```suggestion
    public void saveBitmap(File file, Bitmap mBitmap) {
```

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java [hunk 9]
```diff
@@ -226,8 +237,55 @@ public void onAttach(Activity activity) {
...
+            } else if (requestCode == CAPTURE_IMAGE_ACTIVITY_REQUEST_CODE
+                    && resultCode == RESULT_OK) {
+                uploadImage(clientImageFile);
+
+            } else {
+                Toast.makeText(getActivity(), R.string.havent_picked_image,
+                        Toast.LENGTH_LONG).show();
+            }
+        } catch (Exception e) {
+            Toast.makeText(getActivity(), e.toString(), Toast.LENGTH_LONG)
```
Comment (luckyman20) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java]: Use snackbar instead.

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java [hunk 10]
```diff
@@ -226,8 +237,55 @@ public void onAttach(Activity activity) {
...
+
+            } else if (requestCode == CAPTURE_IMAGE_ACTIVITY_REQUEST_CODE
+                    && resultCode == RESULT_OK) {
+                uploadImage(clientImageFile);
+
+            } else {
+                Toast.makeText(getActivity(), R.string.havent_picked_image,
+                        Toast.LENGTH_LONG).show();
+            }
+        } catch (Exception e) {
```
Comment (luckyman20) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java]:
```suggestion
        } catch (Exception exception) {
```

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java [hunk 11]
```diff
@@ -226,8 +237,55 @@ public void onAttach(Activity activity) {
...
+                imgDecodableString = cursor.getString(columnIndex);
+                cursor.close();
+                Bitmap pickedImage = BitmapFactory.decodeFile(imgDecodableString);
+                saveBitmap(clientImageFile, pickedImage);
+                uploadImage(clientImageFile);
+
+            } else if (requestCode == CAPTURE_IMAGE_ACTIVITY_REQUEST_CODE
+                    && resultCode == RESULT_OK) {
+                uploadImage(clientImageFile);
+
```
Comment (luckyman20) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/clientdetails/ClientDetailsFragment.java]:
```suggestion
```

**Pull request merged**